ultimarc/ui/devices_model.py

```diff
@@ -0,0 +1,177 @@
+    def __init__(self, args, env: (ToolEnvironmentObject, None)):
+        super().__init__()
+
+        self.args = args
+        self.env = env
+        self._device_count = self.env.devices.device_count
+        self._category_ = ''
+        self._ui_dev_info_ = []
+
+        self.setup_info()
+
+    def setup_info(self):
+        """ setup up meta data for the devices and configurations """
+        for dev in self.get_devices():
+            tmp = UIDeviceInfo(name=dev.product_name, class_descr=dev.class_descr,
+                               key=dev.dev_key)
+            tmp.setup_icon(dev.class_id)
+            self._ui_dev_info_.append(tmp)
+
+        # Configuration for non connected devices
+        for device_class in DeviceClassID:
+            tmp = UIDeviceInfo(False, class_descr=device_class.name)
+            tmp.setup_icon(device_class.value)
+            self._ui_dev_info_.append(tmp)
+
+    def get_devices(self):
+        """ Return a list of devices we should show information for. """
+        return self.env.devices.filter()
+
+    def roleNames(self):
+        # TODO: Add device information to role dict
+        class_descr = 'class_descr'.encode('utf-8')
+        class_id = 'class_id'.encode('utf-8')
+        name = 'name'.encode('utf-8')
+        key = 'key'.encode('utf-8')
+        category = 'category'.encode('utf-8')
+        icon = 'icon'.encode('utf-8')
+        connected = 'connected'.encode('utf-8')
+        roles = {
+            DeviceRoles.DEVICE_CLASS: class_descr,
+            DeviceRoles.DEVICE_CLASS_ID: class_id,
+            DeviceRoles.PRODUCT_NAME: name,
+            DeviceRoles.PRODUCT_KEY: key,
+            DeviceRoles.CATEGORY: category,
+            DeviceRoles.ICON: icon,
+            DeviceRoles.CONNECTED: connected
+        }
+
+        return roles
+
+    def rowCount(self, parent):
+        if parent.isValid():
+            return 0
+        return len(self._ui_dev_info_)
+
+    def data(self, index: QModelIndex, role):
+        if not index.isValid():
+            return None
+
+        if role == DeviceRoles.DEVICE_CLASS:
+            i = 0
+            for ui_dev in self._ui_dev_info_:
+                if i == index.row():
+                    return ui_dev.class_descr
+                i = i + 1
+
+        if role == DeviceRoles.DEVICE_CLASS_ID:
+            i = 0
+            for ui_dev in self._ui_dev_info_:
+                if i == index.row():
+                    return ui_dev.class_id
+                i = i + 1
+
+        if role == DeviceRoles.PRODUCT_NAME:
+            i = 0
+            for ui_dev in self._ui_dev_info_:
+                if i == index.row():
+                    return ui_dev.name
+                i = i + 1
+
+        if role == DeviceRoles.PRODUCT_KEY:
+            i = 0
+            for ui_dev in self._ui_dev_info_:
+                if i == index.row():
+                    return ui_dev.key
+                i = i + 1
```

Comment:
All these If statements could be combined into one by creating a lookup dictionary of roles (keys) and ui_dev property names (values).  Then there can be one for loop to get the correct ui_dev object and then use the builtin 'getattr()' function to get the ui_dev property value using the property name found in the role dictionary.

Reply:
I'll refactor this after merge.